### PR TITLE
apply: update keyreg txn handling of StateProofID when going offline

### PR DESCRIPTION
**Pull request by cce**
## Summary

We are not setting record.StateProofID from the values in the keyreg transaction for the case when it is empty (and an account is going offline).

## Test Plan

Updated TestStateProofPKKeyReg to fail without this change.

### CODE REVIEW

**Review by gmalouf — COMMENTED** [review]
Looks right to me, confirmed desired behavior off of `wellFormedStateProofK`

**Review by algoidan — DISMISSED** [review]

File: ledger/apply/keyreg.go [hunk 1]
```diff
@@ -49,6 +49,9 @@ func Keyreg(keyreg transactions.KeyregTxnFields, header transactions.Header, bal
 	// (or, if the voting or selection keys are zero, offline/not-participating)
 	record.VoteID = keyreg.VotePK
 	record.SelectionID = keyreg.SelectionPK
+	if balances.ConsensusParams().EnableStateProofKeyregCheck {
+		record.StateProofID = keyreg.StateProofPK
+	}
```
Comment (winder) [inline ledger/apply/keyreg.go:54]: nit: why not follow the existing pattern and set `record.StateProofID = merklesignature.Verifier{}` below along with the other fields being cleared out?

**Review by algorandskiy — CHANGES_REQUESTED** [review]
The fix is right. But please do not merge it since it will change account hashes => catchpounts.

File: ledger/apply/keyreg.go (hunk 1 above)
Comment (cce) [reply ledger/apply/keyreg.go:54]: I thought this was following the existing pattern more faithfully (see the comment above where it describes the change to the Voting and Selection IDs)

**Comment by cce** [comment]
@algorandskiy I think the ClearOfflineStateProofID consensus param I added will make this safe to merge and release in a future consensus upgrade?

File: ledger/apply/keyreg.go [hunk 2]
```diff
@@ -49,6 +49,9 @@ func Keyreg(keyreg transactions.KeyregTxnFields, header transactions.Header, bal
 	// (or, if the voting or selection keys are zero, offline/not-participating)
 	record.VoteID = keyreg.VotePK
 	record.SelectionID = keyreg.SelectionPK
+	if params := balances.ConsensusParams(); params.EnableStateProofKeyregCheck && params.EnableOnlineAccountCatchpoints {
```
Comment (algoidan) [inline ledger/apply/keyreg.go:52]: why don't you define `params` only once?

Comment (algoidan) [inline ledger/apply/keyreg.go:52]: why don't you define `params` only once?

**Comment by algorandskiy** [comment]
replaced by https://github.com/algorand/go-algorand/pull/4184

**Pull request closed without merging**